vumi_message_store/memory_backend_manager.py

```diff
@@ -0,0 +1,363 @@
+        data_version = riak_object.get_data().get('$VERSION', None)
+        while data_version != store_version:
+            migrator = modelcls.MIGRATOR(
+                modelcls, self, data_version, reverse=True)
+            riak_object = migrator(riak_object).get_riak_object()
+            data_version = riak_object.get_data().get('$VERSION', None)
+        yield riak_object.store()
+        returnValue(modelobj)
+
+    def _internal_load(self, modelcls, key, result=None):
+        """
+        This method gets wrapped in a sync or async decorator in __init__().
+        """
+        assert result is None
+        riak_object = self.riak_object(modelcls, key)
+        yield riak_object.reload()
+        was_migrated = False
+
+        # Run migrators until we have the correct version of the data.
+        while riak_object.get_data() is not None:
+            data_version = riak_object.get_data().get('$VERSION', None)
+            if data_version == modelcls.VERSION:
+                obj = modelcls(self, key, _riak_object=riak_object)
+                obj.was_migrated = was_migrated
+                returnValue(obj)
+            migrator = modelcls.MIGRATOR(modelcls, self, data_version)
+            riak_object = migrator(riak_object).get_riak_object()
+            was_migrated = True
+        returnValue(None)
+
+    def index_keys(self, model, index_name, start_value, end_value=None,
+                   return_terms=None):
+        bucket = self.bucket_for_modelcls(model)
+        return bucket.get_index(
+            index_name, start_value, end_value, return_terms=return_terms)
+
+    def index_keys_page(self, model, index_name, start_value, end_value=None,
+                        return_terms=None, max_results=None,
+                        continuation=None):
+        bucket = self.bucket_for_modelcls(model)
+        return bucket.get_index_page(
+            index_name, start_value, end_value, return_terms=return_terms,
+            max_results=max_results, continuation=continuation)
+
+
+class FakeRiakState(object):
+    """
+    An object to hold fake Riak state.
+    """
+
+    def __init__(self, is_sync):
+        self._buckets = {}
+        self._is_sync = is_sync
+        self._delayed_calls = []
+
+    def teardown(self):
+        for delayed in self._delayed_calls:
+            if not (delayed.cancelled or delayed.called):
+                delayed.cancel()
+
+    def delayed_result(self, result):
+        """
+        If in async mode, return the result with some delay to catch code that
+        doesn't properly wait for the deferred to fire.
+        """
+        if self._is_sync:
+            return result
+
+        # Add some latency to catch things that don't wait on deferreds. We
+        # can't use deferLater() here because we want to keep track of the
+        # delayed call object.
+        d = Deferred()
+        delayed = reactor.callLater(0.002, d.callback, result)
```

Comment:
This code delays firing the deferred we return to make sure that the callers are waiting properly. Without the delay, `foo = manager.load(thing)` and `foo = yield manager.load(thing)` are equivalent and `obj.save()` won't leave the reactor dirty.

Now that I think about it, we should be delaying the whole operation rather than just the result, although that only matters for operations that modify state. (We should do the same in FakeRedis, which is where this code comes from originally.)